lib/internal/test_runner/test.js

```diff
@@ -340,6 +348,58 @@ class TestContext {
       loc: getCallerLocation(),
     });
   }
+
+  waitFor(condition, options = kEmptyObject) {
+    validateFunction(condition, 'condition');
+    validateObject(options, 'options');
+
+    const {
+      interval = 50,
+      timeout = 1000,
+    } = options;
+
+    validateNumber(interval, 'options.interval', 0, TIMEOUT_MAX);
+    validateNumber(timeout, 'options.timeout', 0, TIMEOUT_MAX);
+
+    const { promise, resolve, reject } = PromiseWithResolvers();
+    const noError = Symbol();
+    let cause = noError;
+    let intervalId;
+    let timeoutId;
+    const done = (err, result) => {
+      clearInterval(intervalId);
+      clearTimeout(timeoutId);
+
+      if (err === noError) {
+        resolve(result);
+      } else {
+        reject(err);
+      }
+    };
+
+    timeoutId = setTimeout(() => {
+      // eslint-disable-next-line no-restricted-syntax
+      const err = new Error('waitFor() timed out');
+
+      if (cause !== noError) {
+        err.cause = cause;
+      }
+
+      done(err);
+    }, timeout);
+
+    intervalId = setInterval(async () => {
+      try {
+        const result = await condition();
+
+        done(noError, result);
+      } catch (err) {
+        cause = err;
+      }
+    }, interval);
```

Comment:
Isn't this going to easily call multiple concurrent `condition()` (assuming what it does is async and takes more than `interval` ms to run) ?

Reply:
Yes, that is a possibility. I guess what is the expected behavior in that case? I can think of a handful of different behaviors:

- Detect if the previous check is still executing and skip the current check if so.
- Use a second timer instead of an interval and re-arm the timer if necessary only once the check has finished.
- Error out.

The problem with the first two is that it makes the `interval` option's behavior ambiguous. The third option is just not a good UX IMO.

Reply:
I've pushed a change to address this. I went with option 2.

Reply:
I like option 2, thank you